### PR TITLE
[#1244]Update to LSP4J 0.9.0 Part 2

### PR DESCRIPTION
**Pull request by cdietrich**
[#1244] Update to LSP4J 0.9.0 Part 2
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

This is currently WIP as the change is not on LSP4J master yet.
I also want to mark the deprecated methods for removal in Xtext 2.22 an look for an idea on how to do that (besides writing some sentenses to the release notes)

### CODE REVIEW

**Comment by cdietrich** [comment]
@JanKoehnlein i would also see this to be done but as i dont have time to work on it and i did some overtime shifts to do this PR to be able to keep with the upstream LSP4J changes i would perfer to move further deprecations to 2.22

**Comment by miklossy** [comment]
I agree with @cdietrich . I created a new issue for cleaning up the `xtext.ide.server` package and assigned it to milestone `2.22`.

**Comment by JanKoehnlein** [comment]
@cdietrich I appreciate your work a lot. I could go over the code and mark already deprecated stuff as to be deleted as well, if we have a consensus on the deletion strategy in general. I just want to avoid having deletion sweeps with each release.

**Comment by cdietrich** [comment]
Yes I would prefer to do a "we delete stuff"  not too often

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.xtend [hunk 1]
```diff
@@ -80,13 +81,34 @@ class DocumentSymbolService implements IDocumentSymbolService {
...
 		Document document,
 		XtextResource resource,
 		TextDocumentPositionParams params,
 		IResourceAccess resourceAccess,
 		CancelIndicator cancelIndicator
+	) {
+		if (params instanceof DefinitionParams) {
+			return getDefinitions(document, resource, params as DefinitionParams, resourceAccess, cancelIndicator)
+		}
+		throw new IllegalArgumentException("params is not a HoverParams");
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.xtend]: Exception message suffers from copy'n'paste

File: org.eclipse.xtext.ide/xtend-gen/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.java [hunk 2]
```diff
@@ -102,7 +103,22 @@
...
+  /**
+   * @deprecated please override/call {@link #getDefinitions(Document,XtextResource,DefinitionParams,IResourceAccess,CancelIndicator)} instead.
+   * This method is scheduled to be removed with 2.22.
+   */
+  @Deprecated
   public List<? extends Location> getDefinitions(final Document document, final XtextResource resource, final TextDocumentPositionParams params, final IReferenceFinder.IResourceAccess resourceAccess, final CancelIndicator cancelIndicator) {
+    if ((params instanceof DefinitionParams)) {
+      return this.getDefinitions(document, resource, ((DefinitionParams) params), resourceAccess, cancelIndicator);
+    }
+    throw new IllegalArgumentException("params is not a HoverParams");
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/xtend-gen/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.java]: Exception message

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/occurrences/IDocumentHighlightService.java [hunk 3]
```diff
@@ -33,8 +34,23 @@
...
+	/**
+	 * @deprecated please override/call {@link #getDocumentHighlights(Document, XtextResource, DocumentHighlightParams, CancelIndicator)} instead.
+	 * This method is scheduled to be removed with 2.22.
+	 */
+	@Deprecated//(since="2.21",forRemoval=true)
+	default List<? extends DocumentHighlight> getDocumentHighlights(Document document, XtextResource resource, TextDocumentPositionParams params, CancelIndicator cancelIndicator) {
+		if (params instanceof DocumentHighlightParams) {
+			return getDocumentHighlights(document, resource, (DocumentHighlightParams) params, cancelIndicator);
+		}
+		throw new IllegalArgumentException("params is not a SignatureHelpParams");
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/occurrences/IDocumentHighlightService.java]: Exception message

**Pull request merged**